File: benchmark/fusion/deforestation/page7-2.f
Config NonLinear = true;

Inductive Tree = leaf Int | branch {Tree, Tree};

square = \x: Int. * x x;

squaretr = fix (
  \f: Tree -> Compress Tree. \t: Tree.
  match t with
    leaf w -> leaf (square w)
  | branch {l, r} -> branch {f l, f r}
  end
);

sumtr = fix (
  \f: Tree -> Int. \t: Tree.
  match t with
    leaf w -> w
  | branch {l, r} -> + (f l) (f r)
  end
);

main = \t: Tree. sumtr (squaretr t);

File: ml-benchmark/fusion/deforestation/page7-2.ml
type tree = Leaf of int | Branch of tree * tree

let square x = x * x

let rec squaretr t =
  match t with
  | Leaf w -> Leaf (square w)
  | Branch (l, r) -> Branch (squaretr l, squaretr r)

let rec sumtr t =
  match t with
  | Leaf w -> w
  | Branch (l, r) -> sumtr l + sumtr r

let program t = sumtr (squaretr t)